Adds support for transpose+reshape folding

manupak opened the pull request:
Adds support for transpose+reshape folding

Currently the transposes on inputs are only folded
if they the transpose output immediatedly reaches
a matmul or a conv2d. If they encounter, a reshape
in the middle the folding stops. This will confuse
the gemm-like op.

This commit adds further folding to fold reshapes
i.e. expand/collapse shape ops in between.

closes : https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/862

manupak committed:
Adds support for tranpose+reshape folding

Currently the transposes on inputs are only folded
if they the transpose output immediatedly reaches
a matmul or a conv2d. If they encounter, a reshape
in the middle the folding stops. This will confuse
the gemm-like op.

This commit adds further folding to fold reshapes
i.e. expand/collapse shape ops in between.

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -398,39 +398,112 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
+    return (dims[dims.size() - 2] == 2 && dims[dims.size() - 1] == 1);
+  }
+
+  LogicalResult mergeTransposeWithGemmLikeOp(PatternRewriter &rewriter,
```
Could this be `FailureOr<Value>`?

manupak replied:
no, there is nothing returned here except LogicalResult -- Ive changed others const refs to make it clear.

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -398,39 +398,112 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
+        SmallVector<ReassociationIndices, 4> reassocIndices =
+            op.getReassociationIndices();
+        tensor::CollapseShapeOp newCollapseShapeOp =
+            rewriter.create<tensor::CollapseShapeOp>(op.getLoc(), tInput,
```
Isn't `tInput` an output parameter?

krzysz00 replied:
... Oh. It's both. Could use a comment.

manupak replied:
I've added a comment all are inputs to the function.

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -398,39 +398,112 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
+        SmallVector<int64_t, 4> newOutShape = llvm::to_vector<4>(outShape);
+        // Assumption is shape expansions happen only in higher dimensions.
+        for (size_t i = 0; i < dims.size(); i++) {
+          newOutShape[newOutShape.size() - 1 - i] =
```
What's this supposed to be doing?

krzysz00 replied:
If you're trying to apply the transpose to the output shape, maybe go through the reassociation expressions and map those over?

manupak replied:
Basically, I want to lift the expand shape on top of transpose.

I thought about faithfully following the reassociation expressions but then it would associate expanded dim strictly as before -- which is oddly what I dont want.

E.g. 

Before:
tp:                    m x n ---> n x m      
expand_shape: n x m ---> 1 x n x m
                         reassoc : [0, 1] , [2]

After (what I want it to be with no tp)
new_expand_shape: m x n --> 1 x m x n
                                 reassoc : [0, 1] , [2]  is still the same.
                                 
So reassociation seems identical -- just some dims are swapped. 
Here I just wanted to calculate the outShape of new_expand_shape while using the old reassociation.

I maybe wrong -- but strictly associating with reassociation felt like it would be give something like m x 1 x n which is not what we want here ?

manupak replied:
or I could drop expand shape support -- we dont strictly need it (yet). What I m seeing is expanded shapes coming from clients (that has more dims than the ops require) and then our lowering add reshapes to collapse.

manupak replied:
removed it for now -- it seems confusing and I dont see us exercising the case either (unless in a contrived test) -- we can revisit if needed.

krzysz00 replied:
... ah, that makes more sense. 

I think the cases where we have expand_shape after this all could be stuff like adding a group dimension ... but that's usually done with coordinate transforms.

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -443,46 +516,22 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
                    tInput.getDefiningOp<tosa::MatMulOp>()) {
-      setTranspose(matMulOp, "transpose_c", matmulNonTrivial);
+      LogicalResult res = checkMatMulTransposeValid(matMulOp, dims);
+      if (res.failed()) {
```
You can do `if(failed(checkMatMulTranspose(...)) { return failure(); }` which is somewhat more the style.

krzysz00 replied:
Or `.failed() { return failure(); }`

manupak replied:
I want to propagate the error.

krzysz00 replied:
There isn't an error. `LogicalResult` is a wrapper around `bool`

manupak replied:
right.. need to learn more about diagnostic handling. done.

krzysz00 reviewed (commented):
Minor nits and slight confusion, but the idea seems fine.

Also, can we get a test for convolutions?

manupak committed:
* dropped expand shape support for now
* added conv2d tests.

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -406,31 +412,89 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
+  // with tInput. If there are collapse shapes encountered, the collapse
+  // is applied on the tInput.
+  LogicalResult mergeTransposeWithGemmLikeOp(PatternRewriter &rewriter,
+                                             const Value &tOutput,
```
`Value` is a smart pointer, just pass it by value. That is, any and all uses of `const Value&` can be `Value`

manupak replied:
done.

manupak committed:
* address review feedbacks

krzysz00 reviewed (commented):
Wait, hold on, I'm not sure that a uniform `offset` is right? Don't we want to subtract off the number of dimensions that have been removed at that point in the transpose?

manupak committed:
* generalized transpose dim propogation thru collapses.
* added a test case where lower order dims are collapsed.

manupak commented:
Ok, yea I was taking the luxury of assuming that collapses affect only higher order dims.

Now, I've generalized that a bit to track associations and removed the offset. Now the permuted dims are modified as and when a collapse is encountered and added a test case where lower order dims are collapsed as well (though they are unlikely for us to encounter)

krzysz00 reviewed (approved):
This all seems pretty reasonable to me. I'm trusting that the `sort` call is needed for some reason I don't fully understand right now.
